pkg/agent/nodeportlocal/npl_agent_init.go

```diff
@@ -33,36 +35,128 @@ import (
 	"k8s.io/apimachinery/pkg/watch"
 	clientset "k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
 )
 
 const resyncPeriod = 60 * time.Minute
 
+func cleanupNPLAnnotationForPod(kubClient clientset.Interface, pod *corev1.Pod) error {
+	_, ok := pod.Annotations[k8s.NPLAnnotationStr]
+	if !ok {
+		return nil
+	}
+	delete(pod.Annotations, k8s.NPLAnnotationStr)
+	if _, err := kubClient.CoreV1().Pods(pod.Namespace).Update(context.TODO(), pod, metav1.UpdateOptions{}); err != nil {
+		return fmt.Errorf("unable to update Annotation for Pod %s/%s, %s", pod.Namespace,
+			pod.Name, err.Error())
+	}
+	return nil
+}
+
+func addRulesForNPLPorts(portTable *portcache.PortTable, allNPLPorts []rules.PodNodePort) error {
+	for _, nplPort := range allNPLPorts {
+		portTable.AddUpdateEntry(nplPort.NodePort, nplPort.PodPort, nplPort.PodIP)
+	}
+
+	if err := portTable.PodPortRules.AddAllRules(allNPLPorts); err != nil {
+		return nil
+	}
+	return nil
+}
+
+// getPodsAndGenRules fetches all the Pods on this Node and looks for valid NPL Annotations, if they
+// exist with a valid Node Port, it adds the Node port to the port table and rules. If the Node port
+// is invalid or the NPL Annotation is invalid, the NPL Annotation is removed. The Pod event handlers
+// take care of allocating a new Node port if required.
+func getPodsAndGenRules(kubeClient clientset.Interface, portTable *portcache.PortTable,
+	nodeName string) error {
+	podList, err := kubeClient.CoreV1().Pods(metav1.NamespaceAll).List(context.TODO(), metav1.ListOptions{
+		FieldSelector: "spec.nodeName=" + nodeName,
+	})
+	if err != nil {
+		return fmt.Errorf("error in fetching the Pods for Node %s: %s", nodeName, err.Error())
+	}
+
+	allNPLPorts := []rules.PodNodePort{}
+	for _, pod := range podList.Items {
+		// for each Pod:
+		// check if a valid NPL Annotation exists for this Pod:
+		//   if yes, verifiy validity of the Node port, update the port table and add a rule to the
+		//   rules buffer
+		podCopy := pod.DeepCopy()
+		annotations := pod.GetAnnotations()
+		nplAnnotation, ok := annotations[k8s.NPLAnnotationStr]
+		if !ok {
+			continue
+		}
+		nplData := []k8s.NPLAnnotation{}
+		err := json.Unmarshal([]byte(nplAnnotation), &nplData)
+		if err != nil {
+			// if there's an error in this NPL Annotation, clean it up
+			err := cleanupNPLAnnotationForPod(kubeClient, podCopy)
```

Comment:
Could we do patch here, instead of copying the Pod?

Reply:
About this, do you have any pointers regarding the type of patching we can use here? Since we are removing an annotation here, `StrategicMergePatchType` didn't work and `ApplyPatchType` also failed citing that it was an unsupported type. `JsonPatchType` was also not helpful as it didn't replace the annotations list.

From k8s docs: https://kubernetes.io/docs/tasks/manage-kubernetes-objects/update-api-object-kubectl-patch/, it says if the fields don't have a patch strategy, the default is to replace, however I don't see that with Pod Annotations.

Reply:
I did not find way to delete a single annotation by patching. @tnqn : any idea?
But at least we should copy and patch only annotations? Also if we go this way, you need to handle conflicts. See here: https://github.com/vmware-tanzu/antrea/blob/e61cfbc6f42bd159faa5dd6a0ae04980382caa87/pkg/agent/controller/traceflow/packetin.go#L47

Reply:
> But at least we should copy and patch only annotations?

The problems that I mentioned before regarding specifying the patch strategy were for annotations (and for the whole pod as well). I tried patching only the annotations with `StrategicMergePatchType` and `ApplyPatchType`. They didn't seem to work.

Regarding https://github.com/vmware-tanzu/antrea/blob/e61cfbc6f42bd159faa5dd6a0ae04980382caa87/pkg/agent/controller/traceflow/packetin.go#L47 , it looks like conflicts can happen if multiple agents are updating TraceFlow resource. However, do you think we would hit this scenario if there's only one antrea agent updating the pod, since, each agent works on the pods that are on that corresponding node?

Reply:
Actually I should revert what I said. I did not try it yet, but we used to use json merge path to delete a single annotation. Have you tried it does not work?

Reply:
So I tried the annotation-only patching this time with `JSONPatchType` and it worked. Have updated the PR. Thanks!